packages/langium/src/grammar/type-system/type-collector/types.ts

```diff
@@ -376,6 +388,114 @@ function pushReflectionInfo(node: CompositeGeneratorNode, name: string) {
     node.append('}', NL);
 }
 
+function pushDataTypeReflectionInfo(node: CompositeGeneratorNode, union: UnionType) {
+    const subTypes = Array.from(union.subTypes).map(e => e.name);
+    const strings = collectStringValuesFromDataType(union);
+    const regexes = collectRegexesFromDataType(union);
+    const returnString = createDataTypeCheckerFunctionReturnString(subTypes, strings, regexes);
+    node.append(`export function is${union.name}(item: string): item is ${union.name} {`, NL);
+    node.indent(body => {
+        body.append(returnString + ';', NL);
+    });
+    node.append('}', NL);
+}
+
+function createDataTypeCheckerFunctionReturnString(subTypes: string[], strings: string[], regexes: string[]): string {
```

Comment:
Can we also add a check for `number` (`typeof item === 'number')`), `bigint` (`typeof item === 'bigint'`) and `Date` (`item instanceof Date`)? That would make the checks complete.